### DISCUSSION TITLE
Authentication

### DISCUSSION THREAD

**Discussion by mamparra** [O365/python-o365#749] [Q&A] [answered]
Hi,

Looking for assistance regarding authentication.  When the script is first ran, this authenticates fine, and works for a short time (approximately 45 min).  After this, when i assume the token has expired, unfortunately it does not refresh the token, and i'm unsure of what i'm doing wrong.  

I've gone through and deleted cache pyc files, and the token file, with no luck.  The only way i can guarantee to authenticate is to restart the machine, which made me think its related to the caching.

I've stepped through the authentication method with "Justmycode" on false (vscode), with no luck either.

Windows 10 / server 2012 R2
Python 3.9 64 bit
Credentials with a tenant ID and main resource passed 

Any pointers on what i'm doing wrong / can provide to assist your understanding?

**Comment by mamparra** [comment]
On a side note, absolutely love this package.  The use of the message object, & conv ID's is immense, really appreciate the time put in to create this.

**Comment by alejcas** [comment, marked as answer]
Hi, the token refresh is automatic IF the token has a refresh token available.
For that you need to incorporate the "offline_access" permission.

Maybe you don't add that permission?

**Reply by mamparra** [reply]
Thanks for prompt response :).

Ok cool, i'll have a look at that, i don't think i have that permission enabled.
